src/big/big_decimal.cr

```diff
@@ -0,0 +1,249 @@
+require "big_int"
+
+# A `BigDecimal` represents arbitrary precision decimals.
+#
+# It is internally represented by a pair of `BigInt` and `UInt64`: value and scale.
+# Value contains the actual value, and scale tells the decimal point place.
+# e.g. value=1234, scale=2 => 12.34
+#
+# The general idea and some of the arithmetic algorithms were adapted from
+# the MIT/APACHE -licensed https://github.com/akubera/bigdecimal-rs
+
+class InvalidBigDecimalException < Exception
+end
+
+struct BigDecimal
+  ZERO = BigInt.new(0)
+  TEN  = BigInt.new(10)
+  @@max_div_iterations = 100_u64
+
+  include Comparable(BigDecimal)
+  include Comparable(Int)
+  include Comparable(String)
+
+  private property value : BigInt
+  private property scale : UInt64
+  getter value, scale
+
+  # Sets maximum iterations used in division operation. This implicitly
+  # defines the maximum precision of divisions in cases where the
+  # division is not exact.
+  def self.set_max_div_iterations(new_max_div : Int)
+    @@max_div_iterations = new_max_div.to_u64
+  end
+
+  # Creates a new `BigDecimal` from a `String`.
+  #
+  # Allows only valid number strings with an optional negative sign.
+  def initialize(str : String)
+    # disallow every non-valid string
+    if str !~ /^-?[0-9]+(\.[0-9]+)?$/
+      raise InvalidBigDecimalException.new("Invalid BigDecimal: #{str}")
+    end
+
+    if str.includes?('.')
+      v1, v2 = str.split('.')
+      @value = "#{v1}#{v2}".to_big_i
+      @scale = v2.size.to_u64
+    else
+      @value = str.to_big_i
+      @scale = 0_u64
+    end
+  end
+
+  # from `Int`
+  def initialize(num : Int)
```

Comment:
There should be a far faster way to initialize from integers than via to_s.

Reply:
Oh, whoops. Somebody was lazy :P Will fix.